src/nets/ctc_prefix_score.py

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python
+
+# Copyright 2018 Mitsubishi Electric Research Labs (Takaaki Hori)
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+import numpy as np
+import six
+
+
+class CTCPrefixScore(object):
+
+    def __init__(self, x, blank, eos, xp):
+        ''':param x    : xp array of log_softmax (#frame x #label)
+        :param blank: blank id
+        :param eos  : eos id
+        :param xp   : array module
+        '''
+        self.xp = xp
+        self.logzero = -10000000000.0
+        self.blank = blank
+        self.eos = eos
+        self.input_length = len(x)
+        self.x = x
+
+    def initial_state(self):
+        '''Obtain an initial CTC state
+
+        :return: CTC state
+        '''
+        r = self.xp.full((self.input_length, 2), self.logzero, dtype=np.float32)
+        r[0, 1] = self.x[0, self.blank]
+        for i in six.moves.range(1, self.input_length):
+            r[i, 1] = r[i - 1, 1] + self.x[i, self.blank]
+        return r
+
+    def __call__(self, y, cs, r_prev):
```

Comment:
This part looks different from Algorithm 2 in the paper (probably due to log-level operations and some modifications), and I cannot fully follow it... (I'm sorry)

Reply:
Actually it is a little different. The code computes the scores for a set of next labels simultaneously while Algorithm 2 computes the score only for one next label. I will add some comments on this difference.